mesonbuild/build.py

```diff
@@ -3072,12 +3072,16 @@ class TestSetup:
     env: EnvironmentVariables
     exclude_suites: T.List[str]
 
-def get_sources_string_names(sources, backend):
+
+def get_sources_string_names(sources: T.Sequence[T.Union[str, BuildTarget, CustomTarget, CustomTargetIndex, GeneratedList, File, ExtractedObjects]],
+                             backend: Backend) -> T.List[str]:
     '''
     For the specified list of @sources which can be strings, Files, or targets,
     get all the output basenames.
     '''
-    names = []
+    names: T.List[str] = []
+    if isinstance(sources, str):
+        sources = [sources]
```

Comment:
... which means this logic is incorrect, it does nothing except work around incorrect type information.

Reply:
In fact, CustomTarget already incorrectly says it takes a Sequence as `sources`, then converts it to a `list(sources)` which flatly misbehaves if passed the str that it is allowed to be passed. But by the time we hit get_sources_string_names it is guaranteed to be a list anyway, so I'll spare my heartburn over python's typing model w.r.t. CustomTarget and stick to pointing out that this condition right here is impossible.

Reply:
I know this really bother's you, but this is *the way* to have covariant sequences, and we spend a *lot* of time bending over backwards to not use Sequence when it would save us a ton of pain, just have a look at `Interpreter.source_strings_to_files`, and then all of the uses that don't fit the already substantial number of overloads we have.

But it's really correct for `CustomTarget` to use a Sequence. It's madness to do anything else. Otherwise we'd have to write overloads for a combinatorial explosion of arguments.

The better solution, and the one that I'd like to get to, is to *never* pass string file names past the Interpreter, and to convert them all to File at that point. Because this problem basically exists only because `str` happens to be a valid `Sequence`, and Python doesn't have type subtraction to fix this.

Reply:
No it is NOT correct to use a Sequence. It is correct to use a CovariantIndexableContainerizedHolderOfDiscreteElements, key terms being "covariant" and "container". We want a container type.

A list is a container. A tuple is a container. Custom objects designed to be containers are containers. A string is *typed* as a sequence but it is not in fact a "container type".

It is unfortunate that python typing doesn't support container types. I don't think the solution is to reconvert datatypes back to themselves, especially when it is the WRONG object mode. I would be more comfortable if this were an assert, and use of a `'str'` when a `Sequence[str]` is expected produced an "ERROR: Unhandled python exception" rather than being silently converted to `['s', 't', 'r']`.

It's not even the use of Sequence itself that I object to, it's the use of `list()` on sequences. It is the tail wagging the dog. More importantly, it is a bug magnet. I've tried pointing this out on numerous occasions.